MVPWorkshop/filecoin-solidity-library
# Better way to test

### Feature:

This setup provides a more readable and concise way of testing when compared to present rust tests.

Related to #2 . It enables the use of [hardhat's openzeppelin upgrade plugin](https://docs.openzeppelin.com/upgrades-plugins/1.x/) needed for closing the issue.

Note: It creates a hybrid [foundry-hardhat project](https://hardhat.org/hardhat-runner/plugins/nomicfoundation-hardhat-foundry)

### Setup:

- Clone and start the: [filecoin-localnet](https://github.com/filecoin-project/filecoin-fvm-localnet). For this feature only `lotus` and `lotus-miner` are needed (in the docker-compose.yaml you can comment out other services).
- fetch the changes from this branch and `yarn install`
- Run: `npx hardhat run hh-test/market.sample.test.ts --network localnet`

### How it works (Overview)

It utilises [Lotus CLI](https://lotus.filecoin.io/lotus/manage/lotus-cli/) and [Lotus Miner CLI](https://lotus.filecoin.io/storage-providers/operate/lotus-miner-cli/) to create new accounts that will be used in the hardhat test and to issue additional commands needed to sign messages and change the miner's control address (Note: miner's control addresses is the address that has the privilege to publish deals on the behalf of the provider).

### Test flow

Main file: `./hh-test/market.sample.test.ts`

Scenario:
- Generate all of the accounts needed (deployer (F4/ETH address), anyone (F4/ETH address) and client (F3/BLS address))
     - Note: the provider is created by lotus-miner and stays constant between runs
- Fund the newly created accounts (native FIL balances)
- Deploy all contracts (Market and Helper)
- Set the miner's control address to the Market Contract's one (F4 address and not the ETH equivalent)
- Generate the deal parameters
- Serialise the deal parameters using the Helper contract
  - Note: This is CBOR serialisation, due to many custom solidity types I assume there is no better way to do this
- Sign the serialised deal parameters by the client
- Publish the deal (done by anyone)
- Compare all of the expected vs actual values

### Additional useful material
- Lotus/Lotus Miner CLI is your best friend (completely read those docs)
  -  For example, you [can call a deployed contract's method](https://lotus.filecoin.io/lotus/manage/lotus-cli/#lotus-evm-invoke) directly from the command line:
     - I.e.  `docker exec lotus lotus evm invoke t410f...q <ABI encoded params>` 
       - ABI encoded params (hex) include function signature but not the `0x` prefix
  - Or call the builtin actors using the [send](https://lotus.filecoin.io/lotus/manage/lotus-cli/#lotus-send) command and `--method/value` options and providing **CBOR** encoded params (see [CBOR playground](https://cbor.me/))
- [Differences between different FIL addresses](https://docs.filecoin.io/basics/the-blockchain/addresses)
  - Note: each F1-4 address has the corresponding F0 address
- [Filecoin Complete Specification](https://spec.filecoin.io/)
  - For example you can get methods of builtin actor this way (i.e. [Storage Miner Actor](https://spec.filecoin.io/systems/filecoin_mining/storage_mining/storage_miner_actor/#:~:text=3%3A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20a.ChangeWorkerAddress%2C))
